Add constraint for go-open-service-broker-client

carolynvs opened the pull request:
* Add a dep constraint for the osb-client in preparation for upcoming breaking changes. The constraint is for 0.0.1 <= X < 0.1.0 (patches only).
* Document how to update a dependency with dep in the devguide, including a screencast of me doing this PR.

Closes #1737

MHBauer commented on docs/devguide.md:
```diff
@@ -297,6 +298,36 @@ If you choose third party resources storage, the helm chart will not launch an
+* Gopkg.toml - the dep manifest, this is intended to be hand-edited and contains a set of
+constraints and other rules for dep to apply when selecting appropriate versions of dependencies.
+* Gopkg.lock - the dep lockfile (generated). For the most part, don't edit this file because dep
+completely overwrites that file during `dep ensure`. If you run into trouble with this file, ask.
```
ask who?

MHBauer replied:
Can we summarize the rule?

carolynvs replied:
> ask who?

In the PR or our slack channel? Should I put my phone number in there? 😉 

> Can we summarize the rule?

I wish I could! Let me try.... 🤔 

"You may hand edit Gopkg.lock when to instruct dep to select a specific revision on a branch that is not the HEAD of the branch while continuing to use a branch constraint in Gopkg.toml, which will make it easier to upgrade and track the branch in the future (as opposed to pinning the revision in a constraint)."

But that is such an advanced case that I'm not sure it's worth mentioning. Maybe just stick to "don't edit this file".

nilebox replied:
I would suggest to keep it "don't edit lock file".
The advanced case you described is about developing a branch which we don't have any restrictions for. But once the PR against master is open, we want to keep it reproducible and not allow manual editing.

MHBauer commented on docs/devguide.md:
```diff
@@ -297,6 +298,36 @@ If you choose third party resources storage, the helm chart will not launch an
+review your pull request.
+
+If you use VS Code, we recommend installing the [dep extension](https://marketplace.visualstudio.com/items?itemName=carolynvs.dep).
+It provides snippets and improved highlighting that makes it easier to work with dep.
```
I'll have to look for an emacs plugin...

carolynvs replied:
I actually use Goland, not VS Code, and really wish someone would make a plugin for that for me! 😀 I wimped out and made that dep extension for VS Code instead only because it's so darn easy to write plugins for VS code.

MHBauer replied:
I think @n3wscott mentioned that a few weeks or months ago. Maybe he has a tip?

MHBauer commented on docs/devguide.md:
```diff
@@ -297,6 +298,36 @@ If you choose third party resources storage, the helm chart will not launch an
+Gopkg.toml already, add a constraint for it and set the version.
+1. Run `dep ensure -v` to sync Gopkg.lock and vendor/ with the updated version.
+
+[Watch a screencast](https://youtu.be/yvL66s_hQ94)
```
very nice. I don't recognize the song.

MHBauer replied:
What is the fancy git gui in the end?

carolynvs replied:
That is Atlassian Source Tree

nilebox replied:
Haha nice

carolynvs commented:
@arschles Rebased!

carolynvs commented:
@arschles Rebased for great goodness!

nilebox commented on docs/devguide.md:
```diff
@@ -297,6 +298,36 @@ If you choose third party resources storage, the helm chart will not launch an
+## Dependency Management
+We use [dep](https://golang.github.io/dep) manage our dependencies. We commit the resulting
+vendor directory to ensure repeatable builds and isolation from upstream source disruptions.
+Because vendor is committed, you do not need to interact with dep unless you are
```
Unrelated question: do we run `dep ensure` (and check that the exit code is `0`) as part of the build to make sure that `vendor` directory is up to date with `Gopkg.lock`?

jboyd01 replied:
nit:  We use dep .. *to* manage  (missing the word "to")

carolynvs replied:
Yup! That's handled in the `verify-vendor` target, which is invoked during `verify`.

nilebox commented on docs/devguide.md:
```diff
@@ -297,6 +298,36 @@ If you choose third party resources storage, the helm chart will not launch an
+1. Import the package in the code and use it.
+1. Run `dep ensure -v` to sync Gopkg.lock and vendor/ with your changes.
+
+### Change the version of a dependency
```
Do we have any guidelines on dependency versioning, e.g. is master ok? or do we require major/minor/strict version?

carolynvs replied:
Dep discourages (but doesn't restrict) using a specific revision in the Gopkg.toml file. You can track master (usually for projects that don't version their releases), another branch, or a specific non-semver tag, semver range, etc. I can add a comment about preferring a semver when available, and when not using master instead of a specific revision. If a package is pinned, it should probably be accompanied with a comment explaining why.

I'll update the PR with that.

carolynvs commented:
I have simplified the explanation of Gopkg.lock to "don't edit this file", fixed the typo, and added guidelines for selecting a version for a dependency.
